src/mscorlib/src/System/Random.cs

```diff
@@ -10,205 +10,192 @@
-      private const int MSEED = 161803398;
-      private const int MZ = 0;
-    
-      
-      //
-      // Member Variables
-      //
-      private int inext;
-      private int inextp;
-      private int[] SeedArray = new int[56];
-    
-      //
-      // Public Constants
-      //
-    
-      //
-      // Native Declarations
-      //
-    
-      //
-      // Constructors
-      //
-    
-      public Random() 
-        : this(Environment.TickCount) {
-      }
-    
-      public Random(int Seed) {
-        int ii = 0;
-        int mj, mk;
-    
-        //Initialize our Seed array.
-        int subtraction = (Seed == Int32.MinValue) ? Int32.MaxValue : Math.Abs(Seed);
-        mj = MSEED - subtraction;
-        SeedArray[55]=mj;
-        mk=1;
-        for (int i=1; i<55; i++) {  //Apparently the range [1..55] is special (Knuth) and so we're wasting the 0'th position.
-          if ((ii += 21) >= 55) ii -= 55;
-          SeedArray[ii]=mk;
-          mk = mj - mk;
-          if (mk<0) mk+=MBIG;
-          mj=SeedArray[ii];
+    public class Random
+    {
+        //
+        // Private Constants 
+        //
+        private const int MBIG = Int32.MaxValue;
+        private const ulong MAGIC = 6364136223846793005;
+
+
+        //
+        // Member Variables
+        //
+        private ulong state;
+        private ulong inc;
+
+        //
+        // Public Constants
+        //
+
+        //
+        // Native Declarations
+        //
+
+        //
+        // Constructors
+        //
+
+        public Random()
+          : this(Environment.TickCount)
+        {
+        }
+
+        public Random(int Seed) : this((ulong)Seed)
+        {
+
+        }
+        
+        public Random(ulong LongSeed, ulong OutputSequence = 0)
+        {
+            state = 0;
+            inc = (OutputSequence << 1) | 1;
+            InternalSample();
+            state += LongSeed;
+            InternalSample();
+        }
+
+        //
+        // Package Private Methods
+        //
+
+        /*====================================Sample====================================
+        **Action: Return a new random number [0..1) and reSeed the Seed array.
+        **Returns: A double [0..1)
+        **Arguments: None
+        **Exceptions: None
+        ==============================================================================*/
+        protected virtual double Sample()
+        {
+            //Including this division at the end gives us significantly improved
+            //random number distribution.
+            return (InternalSample() * (1.0 / MBIG));
+        }
+
+        private int InternalSample()
+        {
+            var oldstate = state;
+            state = oldstate * MAGIC + inc;
+            var xorshifted = (uint)(((oldstate >> 18) ^ oldstate) >> 27);
+            var rot = (uint)oldstate >> 59;
+            return (int) ((xorshifted >> (byte)rot) | (xorshifted << (byte)((-rot) & 31)));
+        }
+
```

Comment:
And I do that by adding `[MethodImplOptions.AggressiveInlining]` in `InternalSample`?